### ISSUE TITLE
Better privilege handling

### ISSUE THREAD

**Issue by jjchico** [sandain/MinecraftServerControlScript#91]
## Rationale

Currently, mscs can only be executed by root or a privileged user $USER_NAME (defaults to "minecraft"). Permission are assured by the checkPermissions() method in the script that forces the script to be executed only by root or $USER_NAME and by the execute() method, that executes all critical tasks as $USER_NAME even if the script is executed by root, implementing a kind of privilege dropping solution.
This approach works correctly in general but has a number of limitation in my modest opinion:

1. A dedicated user is always needed to run the script. A competent system administrator my set-up adequate folder and group permissions so that a group of user my control the servers, but this is explicitly blocked by the script.
1. When executed as root, even if critical tasks are handled by the "execute" method, the rest of the script is still executed by root with a risk of severe system damage.
1. Having to pipe all significant work through "execute" is annoying, development time consuming, low performing and ....... ugly? Still it does not eliminates a severe security risk (see above).

## Proposal

My proposal is to add a small wrapper script that sorts out privileges and executes the real control script with the unprivileged user. The new scripts would be:

* mscs: wrapper script named after the original full script for compatibility. It would behave exactly as the current "mscs". It will read $USER_NAME and executes the real script through "sudo" completely unprivileged.
* msctl (or mscs.real, or mscs.user, ...): The current script but made completely user-agnostic. In particular, "execute" and "checkPermissions" could be dropped.

## Benefits

* Solves the current limitations cited above.
* mscs.real could be used if no permission handling is needed or if it is done by the sysadmin. Also to execute unprivileged operations like "list", "status", etc. Ideal for administrators offering MC server hosting: user are free to run mscs.real as their user name. No further admin settings needed.
* More secure: the whole script would be executed unprivileged (no root).
* Better permission control: through "sudo", the sysadmin can control who can execute the script as the MC user, not just $USER_NAME or root.
* Completely transparent migration: "mscs" wrapper will act exactly as the current "mscs".
* Great code simplification in mscs.real by removing "execute".

## Drawbacks

* One more file in the project (5-10 lines of code).

I would be happy to implement it, but I would like to get some opinions from the community (more than one is a community, isn't it?)

**Comment by sandain** [comment]
In general I'm ok with the idea.  I always considered the `execute` method a bit of a hack that makes the script harder to follow (and as you mentioned harder to program with... do I need to escape the quotes in this command or not...).  I just want to make sure that any change we make to the way permissions are handled is at least as secure as the current method, I don't want my server's startup sequence to run the minecraft servers as root.

Perhaps another solution to the problem would be to drop the `checkPermissions` method but keep the `execute` method.  Instead of checking whether the user is `root` or `$USER_NAME`, we could just use the file permissions already in place.  If the `start` command finds that it has RW access to the world directory (because and of the U, G, or O bits are set), it would be allowed to start the server.  If the `ls` command is able to get R access to the world directory, the world can be listed.  This should allow me to do something like `sudo adduser sandain minecraft` so that I can operate the script without using `sudo` or `sudo -u minecraft bash`.

**Comment by jjchico** [comment]
You got the point. The new method would be way more secure than the current approach since you will have the option to execute everything without needing root at all. And in case you want to use root (at start-up or to give privileges to other users) the only code executed as root would be that of the wrapper, that simply executes the real script with an unprivileged user. In fact, the real script may/should include a check to avoid been executed as root, because it would not be necessary now in any case.
Group permission control would be possible completely avoiding root and specific user, but default permissions of the data directory should be carefully set-up, so that, e.g., a world started by a user can be stopped by another user (both in the minecraft group). This will typically involve using ACL's or playing with umask in the script.
But this would be completely compatible with a wrapper forcing the execution with the special user avoiding users to worry about particular file permissions.
I'll make a try and let you know when I have something operational.

**Comment by zanix** [comment]
I'm liking the ideas here

**Comment by jjchico** [comment]
Hi!
I have been doing some tests just using group permissions, so that root could be completely avoided, even at start-up. The only problem is with rdiff-backup, that will always try to preserve ownership and permissions, so that a world backed-up by a user can't not be backed-up again by a different user. This just enforces the use of a wrapper that will assure the script is always executed by a special user.

**Comment by zanix** [comment]
One drawback of `rdiff-backup` is that it has to be ran by the same user every time, but the fact that it does incremental backups I feel outweighs this drawback. You can backup the same world more often and it will use much less space than `tar.gz`.

**Comment by jjchico** [comment]
I have made some progress. You may find the changes in my branch [unpriv-user](https://github.com/jjchico/MinecraftServerControlScript/tree/unpriv-user). Main changes are:

1) Old mscs is now msctl. It can be run by any user and will work in its own account. By default, will create a working dir at $HOME/mscs and can read a per-user config file from $HOME/mscs.conf or $HOME/config/mscs/mscs.conf (more standard). Basically, any user can now do:

    $ msctl create alpha
    $ msctl start alpha
    $ msctl backup alpha
    $ msctl stop alpha
    ...

2) mscs is now a simple wrapper that implements the system-wide behaviour in previous mscs, retaining compatibility.

It is not yet ready for PR because there are modifications to do in Makefile, installation, completion, etc. But I may delay these until we get some testing (I expect your help). msctl and mscs should work.

These are quite a few changes (and some more I have planned) that may go to a 2.0 release including migration instructions.

**Comment by sandain** [comment]
I've run some tests and found some issues:

As my normal user:
```
$ mscs status
Migrating worlds to new mscs.properties config file if needed
Remove file '/opt/mscs/worlds/mscs-worlds-config-migrated' to repeat
the migration process in the next execution.
Minecraft Server Status:
  alpha: running version 1.8.8 (0 of 20 users online).
    Process ID: 1156.
  beta: running version 1.8.8 (0 of 20 users online).
    Process ID: 1492.
  lobby: running.
    Process ID: 2056.
  world: running version 1.8.8 (0 of 20 users online).
    Process ID: 2316.
  gamma: disabled.
```
Here is a list of running processes:
```
$ ps -a -u minecraft
  PID TTY          TIME CMD
 1152 ?        00:00:00 sh
 1154 ?        00:00:00 tail
 1155 ?        00:00:00 sh
 1156 ?        00:02:17 java
 1245 ?        00:00:00 msctl
 1488 ?        00:00:00 sh
 1490 ?        00:00:00 tail
 1491 ?        00:00:00 sh
 1492 ?        00:01:02 java
 1796 ?        00:00:00 msctl
 2047 ?        00:00:00 sh
 2054 ?        00:00:00 tail
 2055 ?        00:00:00 sh
 2056 ?        00:00:06 java
 2312 ?        00:00:00 sh
 2314 ?        00:00:00 tail
 2315 ?        00:00:00 sh
 2316 ?        00:02:44 java
 2523 ?        00:00:00 msctl
 2867 ?        00:00:00 tail
 2868 ?        00:00:00 socat
 2908 ?        00:00:00 tail
 2909 ?        00:00:00 socat
 2984 ?        00:00:00 tail
 2985 ?        00:00:00 socat
 8645 pts/0    00:00:00 ps
```
It appears to be working as expected on my normal user. 

Now lets switch to the `minecraft` user via `su`:
```
$ sudo su -s /bin/bash minecraft
$ echo $HOME
/opt/mscs
$ msctl status
Migrating worlds to new mscs.properties config file if needed
Remove file '/opt/mscs/mscs/worlds/mscs-worlds-config-migrated' to repeat
the migration process in the next execution.
Minecraft Server Status:
```
There are no servers listed and the `mscs-worlds-config-migrated` file was generated in a different directory: `/opt/mscs/mscs-worlds-config-migrated` --> `/opt/mscs/mscs/mscs-worlds-config-migrated`.  These are probably the same issue

Are there any processes running?
```
$ ps -a -u minecraft
  PID TTY          TIME CMD
 1152 ?        00:00:00 sh
 1154 ?        00:00:00 tail
 1155 ?        00:00:00 sh
 1156 ?        00:02:24 java
 1245 ?        00:00:00 msctl
 1488 ?        00:00:00 sh
 1490 ?        00:00:00 tail
 1491 ?        00:00:00 sh
 1492 ?        00:01:05 java
 1796 ?        00:00:00 msctl
 2047 ?        00:00:00 sh
 2054 ?        00:00:00 tail
 2055 ?        00:00:00 sh
 2056 ?        00:00:06 java
 2312 ?        00:00:00 sh
 2314 ?        00:00:00 tail
 2315 ?        00:00:00 sh
 2316 ?        00:02:53 java
 2523 ?        00:00:00 msctl
 2867 ?        00:00:00 tail
 2868 ?        00:00:00 socat
 2908 ?        00:00:00 tail
 2909 ?        00:00:00 socat
 2984 ?        00:00:00 tail
 2985 ?        00:00:00 socat
 8674 pts/0    00:00:00 sudo
 8675 pts/0    00:00:00 su
 8676 pts/0    00:00:00 bash
 8766 pts/0    00:00:00 ps
```
So the servers are still working, the script just can't find them.

Now lets switch to the `minecraft` user via `sudo`:
```
$ sudo -u minecraft bash
$ echo $HOME
/home/jwood
$ msctl status
Migrating worlds to new mscs.properties config file if needed
mkdir: cannot create directory ‘/home/jwood/mscs’: Permission denied
ls: cannot access /home/jwood/mscs/worlds: No such file or directory
ls: cannot access /home/jwood/mscs/worlds: No such file or directory
touch: cannot touch ‘/home/jwood/mscs/worlds/mscs-worlds-config-migrated’: No such file or directory
Remove file '/home/jwood/mscs/worlds/mscs-worlds-config-migrated' to repeat
the migration process in the next execution.
mkdir: cannot create directory ‘/home/jwood/mscs’: Permission denied
ls: cannot access /home/jwood/mscs/worlds: No such file or directory
ls: cannot access /home/jwood/mscs/worlds: No such file or directory
mkdir: cannot create directory ‘/home/jwood/mscs’: Permission denied
ls: cannot access /home/jwood/mscs/worlds: No such file or directory
ls: cannot access /home/jwood/mscs/worlds: No such file or directory
Minecraft Server Status:
```
`$HOME` doesn't get updated using `sudo` this way, resulting in all of these errors.

I guess the question is, is it worth fixing these two use cases?  The issue with the `su` method looks fairly easy to fix since `$HOME` has the correct value, the `sudo` method might be harder to fix.

**Comment by sandain** [comment]
I should add that I ran my tests with the fixes mentioned in issue #95.

**Comment by jjchico** [comment]
Thanks forma testing!

I do not ser any issues here:

1. Msctl is different from mscs. Mscs runs msctl with specific options to behave like previous mscs so you should not mix mscs and msctl. In particular, msctl uses $HOME so that every user can have its own set of servers.

2. sudo -u ... alone does not give you a login shell, but this is not an issue with msctl. Also, there is no clear reason to run msctl after sudo but if you do that you should take care of the environment you inherit as you would do with any other command.

3. Basic rule is msctl is intended to control user servers and mscs to control a system-wide installation.

Sorry for the brevity. I am on the phone.

**Comment by sandain** [comment]
My `minecraft` user is not allowed sudo access (nor should it), so it is unable to utilize the `mscs` script which is why my tests were with the `msctl` script.

Maybe I'm just being difficult.  Your new code makes using my normal user much easier, so there is little reason to switch over to the `minecraft` user to run commands.  However, I think that there are two possible solutions that would keep some backward compatibility with the old script: 1) add a check to the `mscs` script to see if sudo needs to be used, or 2) modify the `msctl` script so that `$LOCATION` doesn't end up as `/opt/mscs/mscs`.  Perhaps the cleanest option is the first, here is an example patch:
```
diff --git a/mscs b/mscs
index c76957f..838472e 100755
--- a/mscs
+++ b/mscs
@@ -16,4 +16,8 @@
 USER_NAME=${USER_NAME:-minecraft}
 LOCATION=${LOCATION:-"/opt/mscs"}
 
-sudo -u $USER_NAME msctl -c /etc/default/mscs -l "$LOCATION" "$@"
+if [ "$USER_NAME" = "$(whoami)" ]; then
+  msctl -c /etc/default/mscs -l "$LOCATION" "$@"
+else
+  sudo -u $USER_NAME msctl -c /etc/default/mscs -l "$LOCATION" "$@"
+fi
```

**Comment by jjchico** [comment]
My idea was that any user in the "minecraft" group could run "mscs" so you completely avoid any root use (except for sudo being set-uid). No need to use the minecraft user directly any more (nor have a password for this user). This can be easily done by sudo configuration like:

    %minecraft		ALL = (minecraft) NOPASSWD: /usr/local/bin/msctl, /usr/bin/msctl

that can be easily added at install time.

The patch will solve the particular case that the minecraft user is used directly, so it is good to have.

My main doubt now is if this branch should replace the current mscs or be a new version. The repo may hold different branches for different versions like:

* mscs-1.0: current version, linked to system-wide "minecraft" user, etc.
* mscs-2.0 (AKA master): new version (this branch), multi-user support, allow further development without being completely compatible with mscs-1.0, migration instructions provided, etc.

My main motivation for a new version is that I plan to improve the options configuration system and it will difficult to do if compatibility with current mscs have to be retained and/or automatic migration provided. I will try to describe the new configuration system in a new issue.

A future 2.0 version may include: the multi-user support in this branch, system-wide operation through the "mscs" wrapper (not necessarily compatible with old mscs) and better options configuration system. So this branch may be a development branch seeking a 2.0 release more than a candidate to replace the current mscs version. This way, users of the 1.0 version may still get bug fixes until they migrate their servers.

If you like the idea, this branch may go to a "dev" branch in the main repo and I will PR on that instead of master.

**Comment by sandain** [comment]
I haven't kept track of versions for the script since its inception.  Maybe I should have.  The script has seen some major changes like this in the past, including a rename from `minecraft_server` to `mscs`.  It has even had code similar to your migrate* methods that I have later removed when I got tired of supporting upgrades.

If everyone thinks that it would be useful to make separate branches and keep track of versions, I'm ok with it, git makes it fairly easy.  I'm not sure that the current version on master should be considered 1.0 though, I would say that version happened in 2011 when I first released the script in this [Minecraft Forum thread](http://www.minecraftforum.net/forums/support/server-support/server-administration/1894314-multi-world-linux-minecraft-server-control-script) and [pastebin](http://pastebin.com/a4uw4cNq).

**Comment by jjchico** [comment]
The main advantage of keeping the current mscs as a separate branch (it may be 1.x or any number you see fit) is that current mscs installations can get bug fixes and delay the migration to a new version with some big changes until these have been carefully tested. It can start by having a separate "dev" branch for the new changes and moving this to "master" when a stable set of changes is ready.
May be tracking versions is not strictly necessary for a small project like this, but it may help user updates, migration code an a possible future packaging.

Best.

**Comment by sandain** [comment]
I went ahead and created a new branch called `version-15.08`, the only patches going to this branch from now on will be bug fixes.  Development will continue to happen on the `master` branch.

**Comment by sandain** [comment]
I accepted your pull request to the `master` branch, then submitted my patch from above.  Closing this issue as complete.  Thanks for all of your hard work on this improvement!

**Comment by zanix** [comment]
This looks great, thanks jjchico.
